diff --git a/roofit/roofitcore/inc/RooAbsAnaConvPdf.h b/roofit/roofitcore/inc/RooAbsAnaConvPdf.h
index 5722ee6..a40f507 100644
--- a/roofit/roofitcore/inc/RooAbsAnaConvPdf.h
+++ b/roofit/roofitcore/inc/RooAbsAnaConvPdf.h
@@ -69,6 +69,8 @@ public:
 
   virtual Bool_t isDirectGenSafe(const RooAbsArg& arg) const ;
     
+  virtual void setCacheAndTrackHints(RooArgSet&) ;
+
 protected:
   Double_t getCoefNorm(Int_t coefIdx, const RooArgSet* nset, const TNamed* rangeName) const ;
 
diff --git a/roofit/roofitcore/src/RooAbsAnaConvPdf.cxx b/roofit/roofitcore/src/RooAbsAnaConvPdf.cxx
index 1812139..603966f 100644
--- a/roofit/roofitcore/src/RooAbsAnaConvPdf.cxx
+++ b/roofit/roofitcore/src/RooAbsAnaConvPdf.cxx
@@ -694,3 +694,16 @@ void RooAbsAnaConvPdf::printMultiline(ostream& os, Int_t contents, Bool_t verbos
 }
 
 
+//_____________________________________________________________________________
+void RooAbsAnaConvPdf::setCacheAndTrackHints(RooArgSet& trackNodes)
+{
+  // Label OK'ed components with cache-and-track
+  RooFIter citer = _convSet.fwdIterator() ;
+  RooAbsArg* carg ;
+  while ((carg=citer.next())) {
+    if (carg->canNodeBeCached()==Always) {
+      trackNodes.add(*carg) ;
+      //cout << "tracking node RooAddPdf component " << carg->IsA()->GetName() << "::" << carg->GetName() << endl ;
+    }
+  }
+}
