diff --git a/roofit/roofitcore/src/RooAbsReal.cxx b/roofit/roofitcore/src/RooAbsReal.cxx
index 9cdd13a..78466fb 100644
--- a/roofit/roofitcore/src/RooAbsReal.cxx
+++ b/roofit/roofitcore/src/RooAbsReal.cxx
@@ -2019,7 +2019,7 @@ RooPlot* RooAbsReal::plotOn(RooPlot *frame, PlotOpt o) const
     }
 
     // add this new curve to the specified plot frame
-    frame->addPlotable(curve, o.drawOptions);
+    frame->addPlotable(curve, o.drawOptions, o.curveInvisible);
 
     if (projDataSel!=o.projData) delete projDataSel ;
        
@@ -3493,6 +3493,14 @@ void RooAbsReal::logEvalError(const char* message, const char* serverValueString
   } else if (_evalErrorMode==CollectErrors) {
     _evalErrorList[this].first = oss2.str().c_str() ;
     _evalErrorList[this].second.push_back(ee) ;
+    if (_evalErrorList[this].second.size() > 2000) {
+       ee = _evalErrorList[this].second.front();
+       coutE(Eval) << "RooAbsReal::logEvalError(" << GetName() << ") evaluation error, " << endl 
+                   << " origin       : " << oss2.str() << endl 
+                   << " message      : " << ee._msg << endl
+                   << " server values: " << ee._srvval << endl ;       
+       _evalErrorList[this].second.pop_front();
+    }
   }
 
   inLogEvalError = kFALSE ;
