diff --git a/roofit/roofitcore/src/RooAbsData.cxx b/roofit/roofitcore/src/RooAbsData.cxx
index af8a193..61cb9f9 100644
--- a/roofit/roofitcore/src/RooAbsData.cxx
+++ b/roofit/roofitcore/src/RooAbsData.cxx
@@ -939,7 +939,7 @@ Double_t RooAbsData::moment(RooRealVar &var, Double_t order, Double_t offset, co
   }
 
   // Check if dataset is not empty
-  if(sumEntries() == 0.) {
+  if(sumEntries(cutSpec, cutRange) == 0.) {
     coutE(InputArguments) << "RooDataSet::moment(" << GetName() << ") WARNING: empty dataset" << endl ;
     return 0;
   }
@@ -961,7 +961,7 @@ Double_t RooAbsData::moment(RooRealVar &var, Double_t order, Double_t offset, co
     
     sum+= weight() * TMath::Power(varPtr->getVal() - offset,order);
   }
-  return sum/sumEntries();
+  return sum/sumEntries(cutSpec, cutRange);
 }
 
 
@@ -998,7 +998,7 @@ Double_t RooAbsData::corrcov(RooRealVar &x,RooRealVar &y, const char* cutSpec, c
   if (!xdata||!ydata) return 0 ;
 
   // Check if dataset is not empty
-  if(sumEntries() == 0.) {
+  if(sumEntries(cutSpec, cutRange) == 0.) {
     coutW(InputArguments) << "RooDataSet::" << (corr?"correlation":"covariance") << "(" << GetName() << ") WARNING: empty dataset, returning zero" << endl ;
     return 0;
   }
@@ -1024,12 +1024,12 @@ Double_t RooAbsData::corrcov(RooRealVar &x,RooRealVar &y, const char* cutSpec, c
   }
 
   // Normalize entries
-  xysum/=sumEntries() ;
-  xsum/=sumEntries() ;
-  ysum/=sumEntries() ;
+  xysum/=sumEntries(cutSpec, cutRange) ;
+  xsum/=sumEntries(cutSpec, cutRange) ;
+  ysum/=sumEntries(cutSpec, cutRange) ;
   if (corr) {
-    x2sum/=sumEntries() ;
-    y2sum/=sumEntries() ;
+    x2sum/=sumEntries(cutSpec, cutRange) ;
+    y2sum/=sumEntries(cutSpec, cutRange) ;
   }
 
   // Cleanup
@@ -1065,7 +1065,7 @@ TMatrixDSym* RooAbsData::corrcovMatrix(const RooArgList& vars, const char* cutSp
 
 
   // Check if dataset is not empty
-  if(sumEntries() == 0.) {
+  if(sumEntries(cutSpec, cutRange) == 0.) {
     coutW(InputArguments) << "RooDataSet::covariance(" << GetName() << ") WARNING: empty dataset, returning zero" << endl ;
     return 0;
   }
@@ -1110,12 +1110,12 @@ TMatrixDSym* RooAbsData::corrcovMatrix(const RooArgList& vars, const char* cutSp
 
   // Normalize sums 
   for (Int_t ix=0 ; ix<varList.getSize() ; ix++) {
-    xsum[ix] /= sumEntries() ;
+    xsum[ix] /= sumEntries(cutSpec, cutRange) ;
     if (corr) {
-      x2sum[ix] /= sumEntries() ;
+      x2sum[ix] /= sumEntries(cutSpec, cutRange) ;
     }
     for (Int_t iy=0 ; iy<varList.getSize() ; iy++) {      
-      xysum(ix,iy) /= sumEntries() ;
+      xysum(ix,iy) /= sumEntries(cutSpec, cutRange) ;
     }
   }    
 
@@ -1203,7 +1203,7 @@ RooRealVar* RooAbsData::rmsVar(RooRealVar &var, const char* cutSpec, const char*
 
   // Fill in this variable's value and error
   Double_t meanVal(moment(var,1,0,cutSpec,cutRange)) ;
-  Double_t N(sumEntries());
+  Double_t N(sumEntries(cutSpec, cutRange));
   Double_t rmsVal= sqrt(moment(var,2,meanVal,cutSpec,cutRange)*N/(N-1));
   rms->setVal(rmsVal) ;
   rms->setError(rmsVal/sqrt(2*N));
